processor/src/main/java/org/mapstruct/ap/internal/model/source/SubClassMappingOptions.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.internal.model.source;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+
+import org.mapstruct.ap.internal.gem.SubClassMappingGem;
+import org.mapstruct.ap.internal.gem.SubClassMappingsGem;
+import org.mapstruct.ap.internal.util.FormattingMessager;
+import org.mapstruct.ap.internal.util.TypeUtils;
+
+import static java.util.Objects.nonNull;
+import static org.mapstruct.ap.internal.util.Message.SUBCLASSMAPPING_ILLEGAL_SUBCLASS;
+import static org.mapstruct.ap.internal.util.Message.SUBCLASSMAPPING_METHOD_SIGNATURE_NOT_SUPPORTED;
+
+/**
+ * Represents a sub class mapping as configured via {@code @SubClassMapping}.
+ *
+ * @author Ben Zegveld
+ */
+public class SubClassMappingOptions extends DelegatingOptions {
+
+    private TypeMirror sourceClass;
+    private TypeMirror targetClass;
+
+    public SubClassMappingOptions(TypeMirror sourceClass, TypeMirror targetClass, DelegatingOptions next) {
+        super( next );
+        this.sourceClass = sourceClass;
+        this.targetClass = targetClass;
+    }
+
+    @Override
+    public boolean hasAnnotation() {
+        return sourceClass != null && targetClass != null;
+    }
+
+    public static List<SubClassMappingOptions> getInstanceOn(ExecutableElement method, MapperOptions mapperOptions,
+                                                       FormattingMessager messager, TypeUtils typeUtils) {
+
+        SubClassMappingsGem subClassMappings = SubClassMappingsGem.instanceOn( method );
+        List<SubClassMappingGem> subClassMappingAnnotations;
+        if ( nonNull( subClassMappings ) ) {
+            subClassMappingAnnotations = subClassMappings.value().get();
+        }
+        else if ( nonNull( SubClassMappingGem.instanceOn( method ) ) ) {
+            subClassMappingAnnotations = Arrays.asList( SubClassMappingGem.instanceOn( method ) );
+        }
+        else {
+            return Collections.emptyList();
+        }
+
```

Comment:
My comment here got lost. Let's do this in a similar way like we are doing it for `@Mapping` and `@Mappings`. We don't need to support meta annotations for now, but the location where we pick them up makes more sense in `MethodRetrieval#getMethodRequiringImplementation`. There you also have access to the source parameters, return type etc. This also answers your other 2 questions in this class I beleive.

Reply:
Made the flow for the `@SubClassMapping` and the `@SubClassMappings` equal to `@Mapping` and `@Mappings` with some refactoring. Added a common base class and two specific classes for the flow. This should also make it easier to add more implementations using the same route in the future. Something for the second review round.

Also added a sorter to solve inheritence ordering issues, the more specific a class the earlier it should be in the list. But not yet implemented it, might be that it will only get a real implementation when someone runs into this as an issue.

Reply:
Yup I saw your refactoring. I like it a lot 😄 .

I've left a comment about the sorter in my last review. I don't think we need it for now.